File: backend/app/services/storage.py
"""
存储抽象层 — 支持本地文件系统 / MinIO

使用方式:
    from app.services.storage import get_storage
    storage = get_storage()
    url = await storage.upload(file_data, filename, content_type)
    await storage.delete(object_key)
"""
import uuid
from abc import ABC, abstractmethod
from dataclasses import dataclass
from io import BytesIO
from pathlib import Path
from typing import Optional

from app.core.config import get_settings

UPLOAD_DIR = Path(__file__).resolve().parent.parent.parent / "uploads"


@dataclass(frozen=True)
class StoredObject:
    """Uploaded object metadata persisted by API callers."""

    url: str
    object_key: str


class BaseStorage(ABC):
    """存储后端抽象基类"""

    @abstractmethod
    async def upload(self, file_data: bytes, filename: str, content_type: str, prefix: str = "blueprints") -> StoredObject:
        """上传文件，返回可访问 URL 和存储 object key。"""
        ...

    @abstractmethod
    async def delete(self, url_or_key: str) -> None:
        """删除文件"""
        ...

    @staticmethod
    def _make_object_key(filename: str, prefix: str = "blueprints") -> str:
        """生成唯一 object key: {prefix}/{uuid}{ext}"""
        ext = Path(filename).suffix or ""
        clean_prefix = prefix.strip("/") or "blueprints"
        return f"{clean_prefix}/{uuid.uuid4().hex[:12]}{ext}"


class LocalStorage(BaseStorage):
    """本地文件系统存储"""

    def __init__(self, upload_dir: Optional[Path] = None):
        self.upload_dir = upload_dir or UPLOAD_DIR
        self.upload_dir.mkdir(parents=True, exist_ok=True)

    async def upload(
        self,
        file_data: bytes,
        filename: str,
        content_type: str = "application/octet-stream",
        prefix: str = "blueprints",
    ) -> StoredObject:
        object_key = self._make_object_key(filename, prefix)
        file_path = self.upload_dir / object_key
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_bytes(file_data)
        return StoredObject(url=f"/uploads/{object_key}", object_key=object_key)

    async def delete(self, url_or_key: str) -> None:
        object_key = _object_key_from_url_or_key(url_or_key)
        file_path = self.upload_dir / object_key
        if file_path.exists():
            file_path.unlink()


class TencentCOSStorage(BaseStorage):
    """Tencent COS object storage."""

    def __init__(self):
        try:
            from qcloud_cos import CosConfig, CosS3Client
        except ImportError as exc:
            raise RuntimeError("Tencent COS storage requires 'cos-python-sdk-v5'. Run pip install -r requirements.txt.") from exc

        settings = get_settings()
        self.bucket = settings.tencent_cos_bucket
        self.region = settings.tencent_cos_region
        self.public_base_url = (
            settings.tencent_cos_public_base_url.rstrip("/")
            or f"https://{self.bucket}.cos.{self.region}.myqcloud.com"
        )
        config = CosConfig(
            Region=self.region,
            SecretId=settings.tencent_cos_secret_id,
            SecretKey=settings.tencent_cos_secret_key,
            Scheme="https",
        )
        self.client = CosS3Client(config)

    async def upload(
        self,
        file_data: bytes,
        filename: str,
        content_type: str = "application/octet-stream",
        prefix: str = "blueprints",
    ) -> StoredObject:
        object_key = self._make_object_key(filename, prefix)
        kwargs = {}
        if content_type == "application/pdf":
            kwargs["ContentDisposition"] = "inline"
        self.client.put_object(
            Bucket=self.bucket,
            Body=BytesIO(file_data),
            Key=object_key,
            ContentType=content_type,
            ACL="public-read",
            **kwargs,
        )
        return StoredObject(url=f"{self.public_base_url}/{object_key}", object_key=object_key)

    async def delete(self, url_or_key: str) -> None:
        object_key = _object_key_from_url_or_key(url_or_key, self.public_base_url)
        self.client.delete_object(Bucket=self.bucket, Key=object_key)


def _object_key_from_url_or_key(url_or_key: str, public_base_url: str | None = None) -> str:
    value = url_or_key.strip()
    if public_base_url and value.startswith(public_base_url.rstrip("/") + "/"):
        return value[len(public_base_url.rstrip("/") + "/"):]
    if value.startswith("/uploads/"):
        return value[len("/uploads/"):]
    return value.lstrip("/")


_storage_instance: Optional[BaseStorage] = None


def get_storage() -> BaseStorage:
    """获取存储实例（单例）。默认使用本地存储。"""
    global _storage_instance
    if _storage_instance is None:
        settings = get_settings()
        if settings.storage_backend == "tencent_cos":
            _storage_instance = TencentCOSStorage()
        else:
            _storage_instance = LocalStorage()
    return _storage_instance

File: backend-go/internal/storage/storage.go
package storage

import (
	"bytes"
	"context"
	"fmt"
	"mime"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/google/uuid"
	"github.com/tencentyun/cos-go-sdk-v5"

	"brickplans/internal/config"
)

// StoredObject is the metadata persisted by callers alongside the DB record.
// ObjectKey is kept server-side only (never returned to the client) and is what
// Delete needs.
type StoredObject struct {
	URL       string
	ObjectKey string
}

type Storage interface {
	Upload(data []byte, filename, contentType, prefix string) (*StoredObject, error)
	Delete(urlOrKey string) error
}

// makeObjectKey builds a server-generated key. The extension is preserved but the
// caller is expected to pass a normalized, safe extension (e.g. ".jpg" / ".pdf");
// the random UUID prefix makes the path unpredictable and not user-controlled.
func makeObjectKey(filename, prefix string) string {
	ext := strings.ToLower(filepath.Ext(filename))
	prefix = strings.Trim(prefix, "/")
	if prefix == "" {
		prefix = "blueprints"
	}
	return fmt.Sprintf("%s/%s%s", prefix, uuid.NewString()[:12], ext)
}

func objectKeyFromURLOrKey(urlOrKey, publicBaseURL string) string {
	v := strings.TrimSpace(urlOrKey)
	if publicBaseURL != "" {
		base := strings.TrimRight(publicBaseURL, "/") + "/"
		if strings.HasPrefix(v, base) {
			return v[len(base):]
		}
	}
	if strings.HasPrefix(v, "/uploads/") {
		return v[len("/uploads/"):]
	}
	return strings.TrimLeft(v, "/")
}

// ── LocalStorage ─────────────────────────────────────

type LocalStorage struct {
	uploadDir  string
	publicBase string
}

func NewLocalStorage(cfg *config.Config) (*LocalStorage, error) {
	if err := os.MkdirAll(cfg.UploadDir, 0o755); err != nil {
		return nil, err
	}
	return &LocalStorage{uploadDir: cfg.UploadDir, publicBase: "/uploads"}, nil
}

func (s *LocalStorage) Upload(data []byte, filename, contentType, prefix string) (*StoredObject, error) {
	key := makeObjectKey(filename, prefix)
	full := filepath.Join(s.uploadDir, filepath.FromSlash(key))
	if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
		return nil, err
	}
	if err := os.WriteFile(full, data, 0o644); err != nil {
		return nil, err
	}
	return &StoredObject{URL: s.publicBase + "/" + key, ObjectKey: key}, nil
}

func (s *LocalStorage) Delete(urlOrKey string) error {
	key := objectKeyFromURLOrKey(urlOrKey, "")
	full := filepath.Join(s.uploadDir, filepath.FromSlash(key))
	if err := os.Remove(full); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}

// ServeFile serves an uploaded file by its URL path (e.g. "/avatars/xxx.jpg").
// PDFs are forced to Content-Disposition: inline; a strict nosniff header is
// applied so the browser honors the declared Content-Type (defeats MIME sniffing
// of uploaded .html/.svg as executable content). Path traversal is rejected.
func (s *LocalStorage) ServeFile(w http.ResponseWriter, r *http.Request, urlPath string) {
	rel := strings.TrimPrefix(urlPath, "/")
	uploadAbs, _ := filepath.Abs(s.uploadDir)
	abs, _ := filepath.Abs(filepath.Join(uploadAbs, filepath.FromSlash(rel)))
	if !strings.HasPrefix(abs, uploadAbs+string(os.PathSeparator)) {
		http.NotFound(w, r)
		return
	}
	st, err := os.Stat(abs)
	if err != nil || st.IsDir() {
		http.NotFound(w, r)
		return
	}
	ext := strings.ToLower(filepath.Ext(abs))
	w.Header().Set("X-Content-Type-Options", "nosniff")
	if ext == ".pdf" {
		w.Header().Set("Content-Disposition", "inline")
		w.Header().Set("Content-Type", "application/pdf")
	} else if ct := mime.TypeByExtension(ext); ct != "" {
		w.Header().Set("Content-Type", ct)
	}
	http.ServeFile(w, r, abs)
}

// ── TencentCOSStorage ────────────────────────────────

type TencentCOSStorage struct {
	client        *cos.Client
	publicBaseURL string
}

func NewTencentCOSStorage(cfg *config.Config) (*TencentCOSStorage, error) {
	base := strings.TrimRight(cfg.TencentCOSPublicBaseURL, "/")
	if base == "" {
		base = fmt.Sprintf("https://%s.cos.%s.myqcloud.com", cfg.TencentCOSBucket, cfg.TencentCOSRegion)
	}
	u, err := url.Parse(base)
	if err != nil {
		return nil, err
	}
	client := cos.NewClient(&cos.BaseURL{BucketURL: u}, &http.Client{
		Transport: &cos.AuthorizationTransport{
			SecretID:  cfg.TencentCOSSecretID,
			SecretKey: cfg.TencentCOSSecretKey,
		},
	})
	return &TencentCOSStorage{client: client, publicBaseURL: base}, nil
}

func (s *TencentCOSStorage) Upload(data []byte, filename, contentType, prefix string) (*StoredObject, error) {
	key := makeObjectKey(filename, prefix)
	opt := &cos.ObjectPutOptions{
		ObjectPutHeaderOptions: &cos.ObjectPutHeaderOptions{
			ContentType:        contentType,
			ContentDisposition: "",
		},
		ACLHeaderOptions: &cos.ACLHeaderOptions{XCosACL: "public-read"},
	}
	if contentType == "application/pdf" {
		opt.ObjectPutHeaderOptions.ContentDisposition = "inline"
	}
	if _, err := s.client.Object.Put(context.Background(), key, bytes.NewReader(data), opt); err != nil {
		return nil, err
	}
	return &StoredObject{URL: s.publicBaseURL + "/" + key, ObjectKey: key}, nil
}

func (s *TencentCOSStorage) Delete(urlOrKey string) error {
	key := objectKeyFromURLOrKey(urlOrKey, s.publicBaseURL)
	_, _ = s.client.Object.Delete(context.Background(), key)
	return nil
}

// ── Singleton ────────────────────────────────────────

var (
	instance Storage
	initOnce sync.Once
	initErr  error
)

// Get returns the singleton storage backend (local or tencent_cos).
func Get(cfg *config.Config) (Storage, error) {
	initOnce.Do(func() {
		if cfg.StorageBackend == "tencent_cos" {
			instance, initErr = NewTencentCOSStorage(cfg)
		} else {
			instance, initErr = NewLocalStorage(cfg)
		}
	})
	return instance, initErr
}

// Reset clears the storage singleton. Intended for tests only.
func Reset() {
	instance = nil
	initErr = nil
	initOnce = sync.Once{}
}
